Add functionalities to get,update, delete facility by id (user), making facility registration request by facility user, and admin handling of those requests

nurgissas opened the pull request:
## What was added
- Add get, update, delete facility by user id methods to userServices
- Corresponding to the above methods changes in controller, router
- Add functionality to request facility registration by facility user (methods, controllers, routes)
- Add admin handling of those requests
- Admin can create facility with request.content
- Admin can get (by authorId or all), delete facility registration request
- Last Edit: bug fix of decline facility registration request

## Related Issues
- Didn't run unit test with existing facility id 
- Tested all facility-registration-request related tests
- But run negative unit tests to endpoints targetting above methods

## Future Works
- Search engine to map services

## Checklist
- [x] I have performed unit tests
- [ ] I have added necessary docs
- [ ] I have checked for potential bugs & issues

cwwojin committed:
docs : add docs for github-pages

- add api docs using markdown, for hosting on github-pages
- set file structures and write pages for index, pages/admin

cwwojin committed:
docs : edit README

cwwojin committed:
docs : edit admin, map, stamp pages

nurgissas committed:
Update facility.md

nurgissas committed:
Fix links to pages md files in index.md

nurgissas committed:
fix get stamp-ruleset rewards by facility ID

cwwojin committed:
docs : minor fixes

cwwojin committed:
Merge branch 'feature/docs_pages' of https://github.com/cwwojin/FORK_backend into feature/docs_pages

cwwojin committed:
Merge branch 'main' into feature/docs_pages

cwwojin committed:
docs : userAPI_01

cwwojin committed:
docs : userAPI_02

cwwojin committed:
docs : edit facilityAPI_01

cwwojin committed:
docs : apply hotfix contents

cwwojin committed:
docs : edit facilityAPI_02

cwwojin committed:
docs : edit TOC

cwwojin committed:
feat : edit login method

- edited loginUser in auth services
- User validation is done with raw password, not bcrypt (temporary)
- login method will create token with necessary payload {id, accountId, userType}
- JWT verification will be done elsewhere (lambda)

cwwojin committed:
Merge branch 'main' into feature/auth

cwwojin committed:
feat : add lambda authorizer

- added authorizer to be run as AWS lambda function
- "./lambda" is not for the back-end, just for lambda
- lambda authorizer will :
    - receive a required header "authorization" which contains a Bearer token
    - perform verification and grant permission for API GW
    - if authorization="guest", client is guest (non-logged in)
    - if authorization SUCCESS, request is forwarded to server
    - 3 headers will be added : {id, accountId, userType} -> to identify user in backend

cwwojin committed:
feat : bcrypt for password

- use bcrypt for user passwords
- "user.password" column in DB will store bcrypt-hash rather than raw data
- edited loginUser to validate with bcrypt

cwwojin committed:
feat : add middleware for userType header check

- requests coming from API GW should have headers
    "userType", "accountId", "id" after successful authorization
- add middleware to validate mandatory header "userType"
    - additionally validate optional header "id", "accountId"

nurgissas committed:
Add functionalities to get,update, delete facility by id

nurgissas committed:
Save before changing branching

nurgissas committed:
Add facility_registration_request functionality and functionality to send facility registration request and etc.

nurgissas committed:
Bug Fix - decline facility registration request

nurgissas committed:
Bug Fix of Uri

cwwojin committed:
feat : new facility views, add isUserFavorite

- add method isUserFavorite to check if a user has a facility favorited
    - "/api/users/favorite/:user/has/:facility"
- edited GET queries for facilityService, mapService to utilize new views in DB
    - GET map locations -> now gets address, opening-hours, preferences, profile-img

cwwojin committed:
Merge pull request #21 from cwwojin/feature/favorites_flag

feat : new facility views, add isUserFavorite method

cwwojin committed:
fix : remove image-URI fields in insert methods

- remove required field 'imgUri', 'logoImgUri', etc.
    - since those methods do not have image uploads.

cwwojin commented:
Great work!
1. `POST /api/facilities/facility-requests` - the `content` should have required field `address` and optional fields `openingHours, menu, preferences, stampRuleset, stampRewards` that include all data to be inserted. validate fields accordingly
2. `GET /api/admin/facility-requests` - OK, but adding optional query parameters `authorId, status` would be nice
3. `GET /api/admin/facility-requests/:id` - OK
4. `POST /api/admin/facility-requests/:id/accept` - insert `preferences, stampRuleset, stampRewards` also
5. `POST /api/admin/facility-requests/:id/decline` - OK
6. `DELETE /api/admin/facility-requests/:id` - OK

Also, you don't need to update `updated_at = now()` -> this will be done by DB

cwwojin commented on myapp/src/services/adminService.js:
```diff
@@ -1,104 +1,166 @@
+      throw { status: 404, message: `Request doesn't exist or is not pending` };
+    }
+
+    const data = JSON.parse(request.content);
```
This will give error if the `content` is an object and not JSONstring

cwwojin committed:
feat : check user permissions by type

- added middleware to check if a requesting user's type is in certain types
    - authorization will set the header 'userType'
    - middleware will use the header for checking
- for methods that should restrict access for certain user types, add middleware for type checking

cwwojin committed:
refact : docker image naming, force rebuild

cwwojin committed:
Merge branch 'feature/auth' into feature/user_permissions

cwwojin commented on myapp/src/services/adminService.js:
```diff
@@ -1,104 +1,166 @@
+        values: [1, id],
+      };
+      const result = await db.query(query);
+
```
Here, also insert the facility-manager relationship
`"insert into manages (user_id, facility_id) values ($1, $2)", [request.author_id, facility.id]`

cwwojin commented on myapp/src/services/facilityService.js:
```diff
@@ -167,6 +165,14 @@ class FacilityService {
 
   async updateFacility(id, data) {
+    const facility = await this.getFacilityById(id);
+    const status = await this.getFacilityRegistrationStatus(facility.authorId);
```
This check will not allow facility update, if the facility is either
- made by admin without the request process, or any facilities that are pre-existent in DB
- request is deleted after being accepted

cwwojin committed:
feat : edit methods getMyFacility, updateMyFacility,
 acceptFacilityRegistrationRequest, updateFacility

- acceptFacilityRegistrationRequest :
    - remove JSON-parse
    - add authorId as the new facility's manager
- updateFacility : remove check of facility registration request status
- userService :
    - changed route names & method names to `myfacility`
    - getMyFacility: get list of all my facilities, use facility_detailed view
    - updateMyFacility :
        - optionally update address, openingHours, menu, preferences, stampRuleset + stampRewards

cwwojin committed:
fix : minor fixes in authMiddleware, loginUser

- allow admin to login method

cwwojin committed:
fix : allow only admin to call createUser API method

cwwojin committed:
feat : add KAIST & facility user registration process

- add user registration process
    - facility user : call registerNewUser to register without verification
- KAIST user : go through email verification to register
    1. registerNewUser : insert user to pending_kaist_user table
    2. send verification email with 6-digit code to KAIST email, and save the code's hash in table
    3. user will enter code in client, which will call verifyKAISTUser
    4. verifyKAISTUser : compare user input code with stored hash
    5. verification success -> insert user into "user" table, delete record from pending_kaist_user
- reSendVerificationMail : helper method to refresh the 6-digit code and resend the mail
    - will store the updated hash in table

cwwojin committed:
fix : change JWT expiration time to 1 day

cwwojin committed:
feat : check duplicate KAIST user before register

- GET /api/users -> add optional query parameter 'email'
- on registration of KAIST user, additionally check if a KAIST user with the same email exists
    - throw 409 if duplicate

cwwojin committed:
Merge pull request #19 from cwwojin/feature/auth

Feature/auth : add login / authorization

cwwojin committed:
Merge branch 'dev' into feature/user_permissions

cwwojin committed:
Merge pull request #22 from cwwojin/feature/user_permissions

Feature/user_permissions : Check user types for all API methods

cwwojin committed:
Merge pull request #25 from cwwojin/feature/user_registration

Feature/user_registration : Add user registration process + email verification

cwwojin committed:
fix : return user info in login payload

cwwojin committed:
Merge pull request #27 from cwwojin:hotfix/login_payload

HOTFIX : return user info in login payload

cwwojin committed:
feat : exclude user password in GET requests

cwwojin committed:
Merge pull request #28 from cwwojin:feature/hide_password

feat : exclude user password in GET requests

nurgissas committed:
Fix content json type and facility request is hangled correctly

nurgissas commented:
POST /api/facilities/facility-requests:
- Added optional fields: address, openingHours, menu, preferences, stampRuleset, stampRewards.
- Validated as objects or arrays.
- Content is treated as object, not string

GET /api/admin/facility-requests:
- Added optional query parameters: authorId, status.

POST /api/admin/facility-requests/:id/accept:
- Added insertion for preferences, stampRuleset, stampRewards.

cwwojin committed:
feat : edit query param name, facility-request format

- facility-request format : stampRewards -> stampRuleset.rewards
    - allow combined insert via createStampRuleset

cwwojin committed:
Merge branch 'feature/user_facility' into feature/user_facility_edit

cwwojin committed:
Merge pull request #23 from cwwojin/feature/user_facility_edit

feature/user_facility_edit : edit methods getMyFacility, updateMyFacility, acceptFacilityRegistrationRequest, updateFacility

cwwojin committed:
feat : edit request validation constraints

- allow facility info > email, url to be optional fields
    - allow value of "", skip email validation

cwwojin commented:
@nurgissas 
My changes :

1. `GET /api/admin/facility-requests` change query parameter name to `authorId -> user` (for consistency with other method)
2. Edit format of `content` in facility request : `stampRuleset: { totalCnt: 10, rewards: []}`
    stampRuleset.rewards is the array with rewards -> to use `facilityService.createStampRuleset` for inserting all
1. Change most fields that use `notEmpty() -> exists().isString()` except for `businessId` Also, allow facility URL and email to be empty string

cwwojin commented:
Great Work ! I will merge this after solving conflict with dev branch

cwwojin committed:
apply prettier on codebase

cwwojin committed:
hotfix : apply prettier on dev

cwwojin committed:
Merge pull request #30 from cwwojin:hotfix/prettier

Hotfix/prettier

cwwojin committed:
Merge branch 'dev' into feature/user_facility

cwwojin committed:
fix : merge conflict fix
